builder/test/copy_pipeline.cc

```diff
@@ -176,13 +182,181 @@ TEST_P(copied_result, pipeline) {
+  int offset_y = std::get<2>(GetParam());
+
+  // Make the pipeline
+  node_context ctx;
+
+  auto in = buffer_expr::make(ctx, "in", 2, sizeof(short));
+  auto out1 = buffer_expr::make(ctx, "out1", 2, sizeof(short));
+  auto out2 = buffer_expr::make(ctx, "out2", 2, sizeof(short));
+
+  auto intm = buffer_expr::make(ctx, "intm", 2, sizeof(short));
+
+  var x(ctx, "x");
+  var y(ctx, "y");
+
+  // In this pipeline, the result is copied to the output. We should just compute the result directly in the output.
+  func stencil = func::make(sum3x3<short>, {{in, {bounds(-1, 1) + x, bounds(-1, 1) + y}}}, {{intm, {x, y}}});
+  func copied1 = func::make_copy({intm, {point(x + offset_x), point(y + offset_y)}}, {out1, {x, y}});
+  func copied2 = func::make_copy({intm, {point(x - offset_x), point(y - offset_y)}}, {out2, {x, y}});
+
+  switch (schedule) {
+  case 0: break;
+  case 1:
+    copied1.loops({y});
+    copied2.loops({y});
+    stencil.compute_root();
+    break;
+  case 2:
+    copied1.loops({y});
+    copied2.loops({y});
+    break;
+  }
+
+  pipeline p = build_pipeline(ctx, {in}, {out1, out2});
+
+  // Run the pipeline.
+  const int W = 20;
+  const int H = 10;
+  buffer<short, 2> in_buf({W + 2, H + 2});
+  in_buf.translate(-1, -1);
+  init_random(in_buf);
+
+  buffer<short, 2> out1_buf({W, H});
+  buffer<short, 2> out2_buf({W, H});
+  out1_buf.translate(-offset_x, -offset_y);
+  out2_buf.translate(offset_x, offset_y);
+  out1_buf.allocate();
+  out2_buf.allocate();
+
+  // Not having span(std::initializer_list<T>) is unfortunate.
+  const raw_buffer* inputs[] = {&in_buf};
+  const raw_buffer* outputs[] = {&out1_buf, &out2_buf};
+  test_context eval_ctx;
+  p.evaluate(inputs, outputs, eval_ctx);
+
+  for (int y = 0; y < H; ++y) {
+    for (int x = 0; x < W; ++x) {
+      int correct = 0;
+      for (int dy = -1; dy <= 1; ++dy) {
+        for (int dx = -1; dx <= 1; ++dx) {
+          correct += in_buf(x + dx, y + dy);
+        }
+      }
+      ASSERT_EQ(correct, out1_buf(x - offset_x, y - offset_y));
+      ASSERT_EQ(correct, out2_buf(x + offset_x, y + offset_y));
     }
   }
 
   ASSERT_EQ(eval_ctx.heap.total_count, 0);
 }
 
+TEST_P(copied_result, two_results_padded) {
+  int schedule = std::get<0>(GetParam());
+  int offset_x = std::get<1>(GetParam());
+  int offset_y = std::get<2>(GetParam());
+
+  // Make the pipeline
+  node_context ctx;
+
+  auto in = buffer_expr::make(ctx, "in", 2, sizeof(short));
+  auto out1 = buffer_expr::make(ctx, "out1", 2, sizeof(short));
+  auto out2 = buffer_expr::make(ctx, "out2", 2, sizeof(short));
+
+  auto intm = buffer_expr::make(ctx, "intm", 2, sizeof(short));
+
+  var x(ctx, "x");
+  var y(ctx, "y");
+
+  std::vector<char> padding = {0, 0};
+
+  box_expr intm_bounds = in->bounds();
+  intm_bounds[0].min += 1;
+  intm_bounds[0].max -= 1;
+  intm_bounds[1].min += 1;
+  intm_bounds[1].max -= 1;
+
+  // In this pipeline, the result is copied to the output. We should just compute the result directly in the output.
```

Comment:
Is this comment correct (especially given that aliasing is off in this test)?

Reply:
Moved this to a different branch for more work.